index.js

```diff
@@ -15,36 +16,64 @@ module.exports = compose
  * @return {Function}
  * @api public
  */
-
-function compose (middleware) {
-  if (!Array.isArray(middleware)) throw new TypeError('Middleware stack must be an array!')
-  for (const fn of middleware) {
-    if (typeof fn !== 'function') throw new TypeError('Middleware must be composed of functions!')
+function compose(middleware) {
+  if (!Array.isArray(middleware)) {
+    throw new TypeError('Middleware stack must be an array!');
+  }
+  if (middleware.some(isNotFunction)) {
+    throw new TypeError('Middleware must be composed of functions!');
   }
 
   /**
    * @param {Object} context
+   * @param {Function} last next handler
    * @return {Promise}
    * @api public
    */
+  return (context, last) => {
+    const localMiddleware = isNotFunction(last) ? middleware : [...middleware, last];
+    const wrapper = (next, handler) => wrapHandler(handler, context, next);
+    const chain = localMiddleware.reduceRight(wrapper, getLastHandler());
+    return chain();
+  }
+}
+
+/**
+ * @return {Function}
+ * @api private
+ */
+function getLastHandler() {
+  let called = false;
+  return () => new Promise((resolve, reject) => {
```

Comment:
Why not use async function?

Reply:
I only refactored the original code from #48 a little bit to see how well it performs. So far I don't really see a reason for this pull to be merged since recursion outperforms `reduceRight` in every version of node I've tested it with. If you really think using reduce can benefit the module, I can refactor it a bit more so that it looks better and uses async in both `getLastHandler` and `wrapHandler`.

Reply:
async functions will outperform promises in a coming V8 version (not sure about source here, I think I read it from the V8 blog - or maybe a patch), so might be worth it when that becomes a patch headline.